### PR TITLE
wasm: fix fail-close streams on VM failure.

### PR DESCRIPTION
**Pull request by mathetake**
Fixes https://github.com/envoyproxy/envoy/issues/14947 and properly closes streams. This commit differentiates `failStream` from `closeStream` where the former is called when a VM fails, and the latter is called via `close_stream` or `grpc_close` by user Wasm codes. Notably, we try to send local response with 503 for http streams as expected by the description of `fail_open` api.
The change here is a little and mostly done in Proxy-Wasm C++ host implementation(https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/155).

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

**Comment by mathetake** [comment]
will add test cases..

**Comment by mathetake** [comment]
Weird failure.. maybe I got stale merge, give me a moment

File: test/extensions/filters/network/wasm/wasm_filter_test.cc
Comment (PiotrSikora) [inline test/extensions/filters/network/wasm/wasm_filter_test.cc:454]: This is blocked on https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/64, I'll add this once it's merged, so feel free to re-assign to me (if there are other changes, but don't waste CI cycle on it).

File: test/extensions/filters/network/wasm/wasm_filter_test.cc
Comment (PiotrSikora) [inline test/extensions/filters/network/wasm/wasm_filter_test.cc:436]: This is blocked on https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/64, I'll add this once it's merged, so feel free to re-assign to me (if there are other changes, but don't waste CI cycle on it).

File: test/extensions/filters/network/wasm/wasm_filter_test.cc
Comment (PiotrSikora) [inline test/extensions/filters/network/wasm/wasm_filter_test.cc:273]: I'm confused, why did it pass before?

**Review by PiotrSikora — APPROVED** [review]
Awesome work, thanks!

File: test/extensions/filters/network/wasm/wasm_filter_test.cc
Comment (mathetake) [reply test/extensions/filters/network/wasm/wasm_filter_test.cc:273]: yeah this is a tricky one. Before we didn't have "post" Vm failure check so `onNewConnection` always returns `Continue` even when the trap happens in VM (this case only Rust panics).

Comment (PiotrSikora) [reply test/extensions/filters/network/wasm/wasm_filter_test.cc:273]: Right, thanks!

**Comment by mathetake** [comment]
kindly ping @envoyproxy/dependency-shepherds PTAL! :)

**Comment by PiotrSikora** [comment]
@moderation PTAL for deps.

Also, is there a way we could automate those dependency checks? We've been stuck many times for a multiple days on what appears to be a manual check to verify that the commit is merged into `master` / `main` branch.

**Review by moderation — APPROVED** [review]

**Review by lizan — APPROVED** [review]

**Pull request merged**
